DirectProgramming/DPC++FPGA/include/streaming_cholesky.hpp

```diff
@@ -0,0 +1,248 @@
+          int raw_latency,  // Read after write latency (in iterations) of
+                            // the triangular loop of this function.
+                            // This value depends on the FPGA target, the
+                            // datatype, the target frequency, etc.
+                            // This value will have to be tuned for optimal
+                            // performance. Refer to the Triangular Loop
+                            // design pattern tutorial.
+                            // In general, find a high value for which the
+                            // compiler is able to achieve an II of 1 and
+                            // go down from there.
+          int pipe_size,    // Number of elements read/write per pipe operation
+                            // to read the input matrix
+          typename AIn,     // A matrix input pipe, receive pipe_size
+                            // elements from the pipe with each read
+          typename LOut     // L matrix output pipe, send one elements to the
+                            // pipe with each write.
+                            // Only lower-left elements of L are
+                            // sent in row order, starting with row 0.
+          >
+struct StreamingCholesky {
+  void operator()() const {
+    // Set the computation type to T or ac_complex<T> depending on the value
+    // of is_complex
+    using TT = std::conditional_t<is_complex, ac_complex<T>, T>;
+
+    constexpr int kColumns = rows;
+
+    // Number of lower-left elements in the L output matrix
+    constexpr int kLMatrixSize = kColumns * (kColumns + 1) / 2;
+
+    // Compute Cholesky decompositions as long as matrices are given as inputs
+    while (1) {
+      // Break memories up to store pipe_size elements per bank
+      constexpr short kBankwidth = pipe_size * sizeof(TT);
+      constexpr unsigned short kNumBanks = rows / pipe_size;
+
+      // When specifying numbanks for a memory, it must be a power of 2.
+      // Unused banks will be automatically optimized away.
+      constexpr short kNumBanksNextPow2 =
+          fpga_tools::Pow2(fpga_tools::CeilLog2(kNumBanks));
+
+      [[intel::numbanks(kNumBanksNextPow2)]]  // NO-FORMAT: Attribute
+      [[intel::bankwidth(kBankwidth)]]        // NO-FORMAT: Attribute
+      [[intel::private_copies(4)]]            // NO-FORMAT: Attribute
+      [[intel::max_replicates(1)]]            // NO-FORMAT: Attribute
+      TT a_load[rows][kColumns];
+
+      // Two copies of L to be able to load two complete rows per iteration
+      // Multiple private copies to be able to overlap multiple loop
+      // iterations
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute[rows][kColumns];
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute_copy[rows][kColumns];
+
+      [[intel::private_copies(2)]]  // NO-FORMAT: Attribute
```

Comment:
Where do 4 and 2 come from?

Reply:
They come from looking at the reports what the concurrency values were and hard setting them here so that the compiler won't be doing strange memory configurations for different sizes